Development practices/Turbo frames.md

```diff
@@ -0,0 +1,315 @@
+        = f.input :last_name
+      .col
+        = f.button :submit
+```
+
+The whole form is wrapped inside the same turbo frame that was used for the index page.
+
+Those are all the changes needed in the views for the inline editing to work.
+
+Once the link to edit the author is clicked, the response provided by the #edit action will have its turbo-frame segment extracted, and the content will replace the frame from where the click originated.
+
+### dom_id helper
+
+The `dom_id` helper can be used to convert the given object into a unique id like this:
+```ruby
+# If the author is persisted and its id is 1:
+dom_id(@author) # => 'author_1'
+
+# If the author is a new record:
+dom_id(Author.new) # => 'new_author'
+
+# optional prefix argument
+dom_id(Author.new, 'prefix') # => 'prefix_new_author'
+```
+
+So we could rewrite the view for the #edit action from the previous example like this:
+```ruby
+= turbo_frame_tag [:edit, dom_id(@author)]
+```
+
+However, since the [`turbo_frame_tag`](https://github.com/hotwired/turbo-rails/blob/ea00f3732e21af9c2156cf74dabe95524b17c361/app/helpers/turbo/frames_helper.rb#L38) helper automatically passes the given object to `dom_id`, this can be further shortened to:
+```ruby
+= turbo_frame_tag :edit, @author
+```
+
+Both of those solutions work correctly and generate the same HTML:
+```Html
+<turbo-frame id='edit_author_1'>
+```
+
+### Controllers
+
+One more thing that needs to be configured is the controller. Since we are wrapping the form in a turbo-frame, we want the validation errors to appear if the object we are creating is invalid.
+
+To do that, the server needs to respond with `422 Unprocessable Entity`, otherwise the whole page might break if there were any validation errors.
+
+Example controller code:
+```ruby
+def edit
+  @author = Author.find(params[:id])
+end
+
+def update
+  @author = Author.find(params[:id])
+
+  if @author.update(author_params)
+    redirect_to authors_path
+  else
+    render :edit, status: :unprocessable_entity
+  end
+end
+```
+
+NOTE: If you are using slim, make sure to name the views like `'index.html.slim'` instead of just `'index.slim'` because otherwise the form submissions won't work correctly!
+
+## Targeting navigation into or out of a frame
+Sometimes we may want the most links to operate within the frame context, but not all of them.
+
+### data-turbo-frame
+Let's say that we also want to be able to add a new author on the index page, without redirecting to a completely new page.
+This time we don't want to replace anything that's visible on the index page.
+We actually want to "append" the form for creating a new author to the page.
+
+If we used the same approach as for the inline editing, we would replace the link for creating a new user with the form. Maybe we don't want to do this because we don't want the form to appear at the same position where the button is, but somewhere else on the page. Luckily it is also done very easily with turbo frames.
```

Comment:
```suggestion
If we used the same approach as for the inline editing, we would replace the link for creating a new user with the form. Maybe we don't want to do this because we don't want the form to appear at the same position where the button is, but somewhere else on the page. Luckily it is also done very easily with frame targeting.
```